### PR TITLE
[FIX] AttentionRefine BUG & Add assets & TYPO in prompt-to-prompt project

### PR DESCRIPTION
**Pull request by Beaconsyh08**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

1. Make the prompt-to-prompt project's demo(~/mmagic/projects/prompt_to_prompt/visualize.ipynb) work
2. Make the prompt-to-prompt AttentionRefine function work

## Modification

**1. ~/mmagic/projects/prompt_to_prompt/assets**

In the demo notebook(~/mmagic/projects/prompt_to_prompt/visualize.ipynb), Part VAE vs DDIM Inversion vs Null-Text Inversion, Section 1. VAE reconstruction as the standard result. The code block loads the image path from [projects/prompt_to_prompt/assets/gnochi_mirror.jpeg]. However, the project doesn't contain the specific folder *assets*. Then, I download the gnochi_mirror image from prompt-to-prompt repository(https://github.com/google/prompt-to-prompt) and add it to the mmagic repository.

**2. ~/mmagic/projects/prompt_to_prompt/models/seq_aligner.py**
If you run the demo code in the above notebook, you will find that the AttentionRefine function doesn't work in the Section Prompt-to-prompt Editing. It shows
> TypeError: list indices must be integers or slices, not tupleTypeError: list indices must be integers or slices, not tuple

Hence, the function *get_matrix* in ~/mmagic/projects/prompt_to_prompt/models/seq_aligner.py should be modified to make the function work.

```python
def get_matrix(size_x, size_y, gap):
    matrix = []
    for i in range(size_x + 1):
        sub_matrix = []
        for j in range(size_y + 1):
            sub_matrix.append(0)
        matrix.append(sub_matrix)
    for j in range(1, size_y + 1):
        matrix[0][j] = j * gap
    for i in range(1, size_x + 1):
        matrix[i][0] = i * gap
    return np.array(matrix)
```

3. ~/mmagic/projects/prompt_to_prompt/visualize.ipynb
TYPO "1. VAE reconstruction as the a standard result" --> "1. VAE reconstruction as the standard result"

## BC-breaking (Optional)
N/A
## Use cases (Optional)

```python
from mmengine import MODELS, Config
from mmengine.registry import init_default_scope

init_default_scope('mmagic')
config = 'configs/stable_diffusion/stable-diffusion_ddim_denoisingunet.py'
config = Config.fromfile(config).copy()
StableDiffuser = MODELS.build(config.model)
StableDiffuser = StableDiffuser.to('cuda')

prompts = ["a photo of Eiffel Tower",
           "a photo of Eiffel Tower at night"]
from models.ptp import AttentionRefine
controller = AttentionRefine(prompts, 50, cross_replace_steps=1.0,
                             self_replace_steps=.4, model=StableDiffuser)
images, _ = text2image_ldm_stable(StableDiffuser, prompts, controller, latent=None)
ptp_utils.view_images(images)
```

## Checklist

Submitting this pull request means that,

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmagic/blob/main/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmagic/blob/main/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by OpenMMLab-Assistant-004** [comment]
Hi @Beaconsyh08,

We'd like to express our appreciation for your valuable contributions to the mmagic. Your efforts have significantly aided in enhancing the project's quality.
It is our pleasure to invite you to join our community thorugh Discord_Special Interest Group (SIG) channel. This is a great place to share your experiences, discuss ideas, and connect with other like-minded people. To become a part of the SIG channel, send a message to the moderator, OpenMMLab, briefly introduce yourself and mention your open-source contributions in the #introductions channel. Our team will gladly facilitate your entry. We eagerly await your presence. Please follow this link to join us: ​https://discord.gg/UjgXkPWNqA.

If you're on WeChat, we'd also love for you to join our community there. Just add our assistant using the WeChat ID: openmmlabwx. When sending the friend request, remember to include the remark "mmsig + Github ID".

Thanks again for your awesome contribution, and we're excited to have you as part of our community!

File: projects/prompt_to_prompt/assets/gnochi_mirror.jpeg
Comment (liuwenran) [inline projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]: Hi, looks like this jpeg is not used in your changed files, so this file would be deleted.

Comment (Beaconsyh08) [reply projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]: Thank you for your review! 
The image in question was not used in the code I recently modified. However, it should be noted that this image was indeed featured in the demo code(~/mmagic/projects/prompt_to_prompt/visualize.ipynb). It is possible that this detail may have inadvertently escaped your notice. Alternatively, if there exists a specific rationale for the omission of this image, I would greatly appreciate your elucidation on the matter. 

`from models import ptp_utils
import torch

image_path = 'projects/prompt_to_prompt/assets/gnochi_mirror.jpeg'
prompt = "a cat sitting next to a mirror"
# using vae to reconstruct the image
image_tensor = ptp_utils.load_512(image_path).to('cuda')
vae = StableDiffuser.vae
with torch.no_grad():
    vae_rec = vae(image_tensor, return_dict=False)[0]
ptp_utils.view_images(ptp_utils.tensor_to_nparray(vae_rec))`

Comment (liuwenran) [reply projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]: Placing the image in the code repository would increase the size of the repository. Therefore, we generally adopt the approach of downloading the image from a URL.

We usually do it like this:
```python
import mmcv
mmcv.imread('https://github.com/open-mmlab/mmagic/assets/12782558/c851dcf8-bcfa-424f-9eeb-d01a48ea7127')
```

You could modify the code in visualize.ipynb like this, thank you very much.

Comment (Beaconsyh08) [reply projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]: Thank有for

Comment (Beaconsyh08) [reply projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]:
```python
from models import ptp_utils
import torch
import mmcv

image = mmcv.imread('https://github.com/open-mmlab/mmagic/assets/12782558/c851dcf8-bcfa-424f-9eeb-d01a48ea7127', channel_order='rgb')
prompt = "a cat sitting next to a mirror"
# using vae to reconstruct the image
image_tensor = ptp_utils.load_512(image).to('cuda')
vae = StableDiffuser.vae
with torch.no_grad():
    vae_rec = vae(image_tensor, return_dict=False)[0]
ptp_utils.view_images(ptp_utils.tensor_to_nparray(vae_rec))
···

Comment (Beaconsyh08) [reply projects/prompt_to_prompt/assets/gnochi_mirror.jpeg:1]: Thank you for your explanation. I have made adjustments to the code using the `mmcv.imread` function. The default loading format is set to `bgr`, but in order to align the input with the prompt-to-prompt code, I have set the loading channel_order to be `rgb`. 
```python
from models import ptp_utils
import torch
import mmcv

image = mmcv.imread('https://github.com/open-mmlab/mmagic/assets/12782558/c851dcf8-bcfa-424f-9eeb-d01a48ea7127', 
                    channel_order='rgb')
prompt = "a cat sitting next to a mirror"
# using vae to reconstruct the image
image_tensor = ptp_utils.load_512(image).to('cuda')
vae = StableDiffuser.vae
with torch.no_grad():
    vae_rec = vae(image_tensor, return_dict=False)[0]
ptp_utils.view_images(ptp_utils.tensor_to_nparray(vae_rec))
```
Your review is appreciated.

**Review by liuwenran — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**